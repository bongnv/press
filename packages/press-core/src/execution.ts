import path from "path";
import type { Hook } from "tapable";
import { AsyncParallelHook, AsyncSeriesHook } from "tapable";
import WebpackConfig from "webpack-chain";
import { Options } from "html-minifier";

export type PluginOptions = undefined | Record<string, any>;
export type Plugin = (execution: Execution, options: PluginOptions) => void;
type PluginNoOptions = string | Plugin;
export type Plugable =
  | [PluginNoOptions, PluginOptions]
  | [PluginNoOptions]
  | PluginNoOptions;

export interface Config {
  baseDir?: string;
  plugins?: Plugable[];
}

interface ExecutionHooks {
  prepare: Hook;
  configWebpack: Hook;
  bundle: Hook;
  generate: Hook;
}

interface ExecutionCommands {
  dev: Hook;
  build: Hook;
}

export class Execution {
  baseDir: string;
  outputDir: string;
  isProd: boolean;
  publicPath: string;
  temporaryPath: string;
  serverPath: string;
  htmlMinifierOptions?: Options;
  enhancers: string[];

  commands: ExecutionCommands;
  hooks: ExecutionHooks;

  clientWebpackConfig: WebpackConfig;
  serverWebpackConfig: WebpackConfig;

  urlsToRender: string[];

  constructor(config: Config) {
    this.baseDir = config.baseDir || process.cwd();
    this.outputDir = path.join(this.baseDir, "dist");
    this.isProd = process.env.NODE_ENv === "production";
    this.publicPath = "/";
    this.temporaryPath = path.join(this.baseDir, ".press");
    this.serverPath = path.join(this.temporaryPath, "server");

    const params = ["execution"];

    this.commands = {
      dev: new AsyncSeriesHook(params),
      build: new AsyncSeriesHook(params),
    };

    this.hooks = {
      prepare: new AsyncParallelHook(params),
      configWebpack: new AsyncSeriesHook(params),
      bundle: new AsyncParallelHook(params),
      generate: new AsyncParallelHook(params),
    };

    this.enhancers = [];
    this.clientWebpackConfig = new WebpackConfig();
    this.serverWebpackConfig = new WebpackConfig();
    this.urlsToRender = [];
  }
}
