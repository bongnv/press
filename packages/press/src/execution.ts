import type { Hook } from "tapable";
import { AsyncParallelHook, AsyncSeriesHook } from "tapable";
import WebpackConfig from "webpack-chain";

import type { Config } from "./load-config";
import { Core } from "./plugins/core";
import { DevServer } from "./plugins/dev-server";
import { StaticGen } from "./plugins/static-gen";

interface Plugin {
  // TODO: we may support async apply
  apply(execution: Execution): void;
}

interface ExecutionHooks {
  prepare: Hook;
  configWebpack: Hook;
  bundle: Hook;
  generate: Hook;
}

interface ExecutionCommands {
  dev: Hook;
  build: Hook;
}

export class Execution {
  config: Config;
  commands: ExecutionCommands;
  hooks: ExecutionHooks;

  clientWebpackConfig: WebpackConfig;
  serverWebpackConfig: WebpackConfig;

  urlsToRender: string[];

  constructor(config: Config) {
    this.config = config;
    this.urlsToRender = [];

    // setup hooks
    const params = ["execution"];
    this.commands = {
      dev: new AsyncSeriesHook(params),
      build: new AsyncSeriesHook(params),
    };

    this.hooks = {
      prepare: new AsyncParallelHook(params),
      configWebpack: new AsyncSeriesHook(params),
      bundle: new AsyncParallelHook(params),
      generate: new AsyncParallelHook(params),
    };

    this.clientWebpackConfig = new WebpackConfig();
    this.serverWebpackConfig = new WebpackConfig();
  }

  async execBuild(): Promise<void> {
    this._applyPlugins();
    await this.commands.build.promise(this);
    await this.hooks.prepare.promise(this);
    await this.hooks.configWebpack.promise(this);
    await this.hooks.bundle.promise(this);
    await this.hooks.generate.promise(this);
  }

  async execDev(): Promise<void> {
    this._applyPlugins();
    await this.commands.dev.promise(this);
    await this.hooks.prepare.promise(this);
    await this.hooks.configWebpack.promise(this);
    await this.hooks.bundle.promise(this);
  }

  private _applyPlugins(plugins: Plugin[] = []) {
    plugins.unshift(new Core(), new DevServer(), new StaticGen());
    for (const plugin of plugins) {
      plugin.apply(this);
    }
  }
}
