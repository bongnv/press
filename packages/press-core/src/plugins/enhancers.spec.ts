import path from "path";

import enhancersPlugin from "./enhancers";
import { Execution } from "../execution";

test("Enhancers should apply configWebpack properly", async () => {
  jest
    .spyOn(path, "resolve")
    .mockImplementationOnce((_, localPath) => path.join("/", localPath));

  const execution = new Execution({
    baseDir: "/",
  });

  enhancersPlugin(execution);
  await execution.hooks.configWebpack.promise(execution);
  expect(execution.clientWebpackConfig.toConfig()).toMatchSnapshot();
  expect(execution.serverWebpackConfig.toConfig()).toMatchSnapshot();
});
