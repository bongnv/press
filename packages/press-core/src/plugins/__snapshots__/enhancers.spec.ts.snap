// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Enhancers should apply configWebpack properly 1`] = `
Object {
  "module": Object {
    "rules": Array [
      Object {
        "test": "/vue-app/enhancers-loader.js",
        "use": Array [
          Object {
            "loader": "val-loader",
            "options": Object {
              "enhancers": Array [],
            },
          },
        ],
      },
    ],
  },
}
`;

exports[`Enhancers should apply configWebpack properly 2`] = `
Object {
  "module": Object {
    "rules": Array [
      Object {
        "test": "/vue-app/enhancers-loader.js",
        "use": Array [
          Object {
            "loader": "val-loader",
            "options": Object {
              "enhancers": Array [],
            },
          },
        ],
      },
    ],
  },
}
`;
